File: ml_service/src/ml_service/policy.py
from __future__ import annotations
from dataclasses import dataclass
from ml_service.schemas import EncodingHint, RecommendRequest


class AdaptivePolicy:
    name: str

    def recommend(self, req: RecommendRequest) -> EncodingHint:
        raise NotImplementedError


@dataclass(frozen=True)
class HeuristicPolicy(AdaptivePolicy):
   

    name: str = "heuristic-v1"

    def recommend(self, req: RecommendRequest) -> EncodingHint:
        stats = req.stats
        importance = float(req.importance_score or 0.0)

        bw_bps = (
            stats.available_outgoing_bitrate_bps if stats.available_outgoing_bitrate_bps is not None
            else (stats.send_bitrate_bps if stats.send_bitrate_bps is not None else 1200000.0) # 1.2Mbps 
        )

        
        bw_kbps = max(0.0, bw_bps / 1000.0)
        base_bps = int(max(150000, min(5000000, bw_bps * 0.85))) #make base_bps b/w (0.15Mbps, 5Mbps)

       
        boost = 0.80 + 0.50 * importance  
        target_bps = int(max(150_000, min(6_000_000, base_bps * boost))) #make target_bps b/w (0.15Mbps,6Mbps)

        loss = stats.packet_loss_pct
        if loss is not None:
            if loss > 8:
                target_bps = int(target_bps * 0.65)
            elif loss > 4:
                target_bps = int(target_bps * 0.8)

        
        rtt = stats.rtt_ms
        if rtt is not None and rtt > 250:
            target_bps = int(target_bps * 0.85)

        if bw_kbps < 350:
            scale = 2.0
            fps = 12
        elif bw_kbps < 700:
            scale = 1.5
            fps = 15
        elif bw_kbps < 1300:
            scale = 1.0
            fps = 24
        else:
            scale = 1.0
            fps = 30

        # If importance is high in poor bandwidth preserve resolution drop fps.
        if importance > 0.8 and bw_kbps < 700:
            scale = 1.0
            fps = 12

        return EncodingHint(
            suggested_max_bitrate_bps=target_bps,
            suggested_scale_down_by=scale,
            suggested_fps=fps,
        )

File: frontend/src/adaptive/policy.ts
import type { EncodingHint, WebRtcStats } from "./types";

function clamp(n: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, n));
}

export function heuristicRecommend(stats: WebRtcStats, importanceScore = 0): EncodingHint {
  const importance = importanceScore;

  const bwBps =
    stats.availableOutgoingBitrateBps ?? stats.sendBitrateBps ?? 1_200_000;

  const bwKbps = Math.max(0, bwBps / 1000);

  const baseBps = clamp(bwBps * 0.85, 150_000, 5_000_000);
  const boost = 0.8 + 0.5 * importance; 
  let targetBps = clamp(baseBps * boost, 150_000, 6_000_000);

  const loss = stats.packetLossPct;
  if (loss != null) {
    if (loss > 8) targetBps *= 0.65;
    else if (loss > 4) targetBps *= 0.8;
  }

  const rtt = stats.rttMs;
  if (rtt != null && rtt > 250) targetBps *= 0.85;

  let scale = 1.0;
  let fps = 30;

  if (bwKbps < 350) {
    scale = 2.0;
    fps = 12;
  } else if (bwKbps < 700) {
    scale = 1.5;
    fps = 15;
  } else if (bwKbps < 1300) {
    scale = 1.0;
    fps = 24;
  }

  if (importance > 0.8 && bwKbps < 700) {
    scale = 1.0;
    fps = 12;
  }

  return {
    mode: "webrtc",
    suggested_max_bitrate_bps: Math.round(clamp(targetBps, 150_000, 6_000_000)),
    suggested_scale_down_by: scale,
    suggested_fps: fps,
  };
}
